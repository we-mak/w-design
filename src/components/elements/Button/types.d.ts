import {
  GlobProps,
  SizeProps,
  LinkProps,
  ActionProps,
  SyntheticEventProps,
  MouseEventProps
} from "../../../typings";

export interface ButtonProps
  extends GlobProps,
    SizeProps,
    LinkProps,
    ActionProps,
    SyntheticEventProps,
    MouseEventProps {
  /** button apperance type */
  appearance?:
    | "default"
    | "primary"
    | "danger"
    | "warning"
    | "success"
    | "subtle"
    | "link"
    | "help"
    | "clean";
  /** WAI-ARIA support */
  ariaControls?: string;
  ariaExpanded?: boolean;
  ariaLabel?: string;
  ariaHaspopup?: boolean;
  /** button type attribute*/
  type?: "button" | "submit" | "reset";
  /** Name property of a linked form that the button submits when clicked. */
  form?: string;
  /** selected button */
  isSelected?: boolean;
  /** children */
  children?: any;
}
