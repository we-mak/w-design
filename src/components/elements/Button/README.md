| Props        | Type   | default   | description             |
| ------------ | ------ | --------- | ----------------------- |
| appearance   | string | "default" | applied style of button |
| buttonRef    |        |           |                         |
| type         |        |           |                         |
| size         |        |           |                         |
| fluid        |        |           |                         |
| iconBefore   |        |           |                         |
| iconAfter    |        |           |                         |
| iconOnly     |        |           |                         |
| fluid        |        |           |                         |
| form         |        |           |                         |
| isLoading    |        |           |                         |
| isDisabled   |        |           |                         |
| isSelected   |        |           |                         |
| href         |        |           |                         |
| target       |        |           |                         |
| autoFocus    |        |           |                         |
| onClick      |        |           |                         |
| tabIndex     |        |           |                         |
| children     |        |           |                         |
| ariaHaspopup |        |           |                         |
| ariaExpanded |        |           |                         |
| ariaControls |        |           |                         |
| ariaLabel    |        |           |                         |

- `appearance` in cluded "default", "primary", "danger", "warning", "success", "subtle", "link", "help", "clean"
