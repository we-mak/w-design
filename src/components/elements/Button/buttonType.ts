import {
  SyntheticEventProps,
  GlobProps,
  LinkProps,
  FormActionProps,
  SizeProps
} from "../../../common/propsType";

export interface BtnProps
  extends GlobProps,
    SyntheticEventProps,
    LinkProps,
    FormActionProps,
    SizeProps {
  /** custom component*/
  customComponent?: string;
  /** button apperance type */
  appearance?:
    | "default"
    | "primary"
    | "danger"
    | "warning"
    | "success"
    | "link";
  /** html button aria-controls*/
  ariaControls?: string;
  /** html button aria-expanded*/
  ariaExpanded?: boolean;
  /** html button aria-label*/
  ariaLabel?: string;
  /** html button aria-haspopup*/
  ariaHaspopup?: boolean;
  /** button ref*/
  buttonRef?: (ref: HTMLElement) => {};
  /** children */
  children?: React.ReactChild | string;
  /** button type attribute*/
  type?: "button" | "submit" | "reset";
  /** Name property of a linked form that the button submits when clicked. */
  form?: string;
  /** selected button */
  isSelected?: boolean;
}
