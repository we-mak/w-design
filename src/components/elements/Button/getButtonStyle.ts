import { css } from "../../../utils/styled";
import { getColorFromTheme } from "../../../utils/themeUtils";
import {
  padding,
  height,
  background,
  backgroundHover,
  backgroundActive,
  backgroundFocus,
  border,
  borderHover,
  borderActive,
  borderFocus,
  text,
  boxShadow
} from "../../../common/themed";
import { fontSizes, radii, colors } from "../../../common/theme";
import { SpinnerStyle } from "../Spinner/Spinner";
/**
 * getAppearanceProps
 * @param appearance appearance style
 * @param colors init colors
 * @param themeColors get colors from provided theme props
 * @returns style applies to each button appearance
 */
const getAppearanceProps = (
  appearance: string,
  colors: object,
  themeColors?: object
) => (themeColors ? themeColors[appearance] : colors[appearance]);

const getFontSizeProps = (
  buttonSizes: "sm" | "md" | "lg",
  fontSizes: Array<string>,
  themeFontSizes?: Array<string>
) => {
  const getTheme = (theme: Array<string>) => {
    switch (buttonSizes) {
      case "sm":
        return theme[0];
      case "md":
        return theme[1];
      case "lg":
        return theme[2];
    }
  };
  if (themeFontSizes) return getTheme(themeFontSizes);

  return getTheme(fontSizes);
};

const getSize = (styleProperty: object, buttonSize: "sm" | "md" | "lg") =>
  styleProperty[buttonSize];

/** Apply style to each appearance when loading props is true */
const getLoadingState = (props: any) => {
  const { isLoading, appearance, theme } = props;
  if (isLoading) {
    if (appearance === "default" || appearance === "link") {
      return SpinnerStyle;
    }

    return css`
      ${SpinnerStyle};
      &::after {
        border-color: ${getColorFromTheme(colors, "WHITE", theme)};
        border-right-color: transparent;
        border-top-color: transparent;
      }
    `;
  }

  return;
};

const getSelected = (props: any) => {
  const { appearance, isSelected, theme } = props;
  if (isSelected) {
    if (
      appearance === "default" ||
      appearance === "link" ||
      appearance === "primary"
    ) {
      return css`
        background: ${getColorFromTheme(colors, "B70", theme)};
        border-color: ${getColorFromTheme(colors, "B50", theme)};
        color: ${getColorFromTheme(colors, "N1", theme)};
        &:hover,
        &:focus,
        &:active {
          background: ${getColorFromTheme(colors, "B70", theme)};
          border-color: ${getColorFromTheme(colors, "B50", theme)};
          color: ${getColorFromTheme(colors, "N1", theme)};
          outline: none;
          box-shadow: none;
        }
      `;
    }

    if (appearance === "danger") {
      return css`
        border-color: ${getColorFromTheme(colors, "R40", theme)};
        background: ${getColorFromTheme(colors, "R30", theme)};
        &:hover,
        &:focus,
        &:active {
          box-shadow: none;
          border-color: ${getColorFromTheme(colors, "R40", theme)};
          background: ${getColorFromTheme(colors, "R30", theme)};
        }
      `;
    }

    if (appearance === "warning") {
      return css`
        border-color: ${getColorFromTheme(colors, "O50", theme)};
        background: ${getColorFromTheme(colors, "O40", theme)};
        &:hover,
        &:focus,
        &:active {
          box-shadow: none;
          border-color: ${getColorFromTheme(colors, "O50", theme)};
          background: ${getColorFromTheme(colors, "O40", theme)};
        }
      `;
    }
  }
  return;
};

/**
 * getButtonStyle
 * Resolves styles based on props
 * @param props button properties
 */
export const getButtonStyle = (props: any) => {
  const { theme, appearance, elementSize, fluid } = props;
  /** Button appearance style using themed*/
  // background
  const bgDefault = getAppearanceProps(
    appearance,
    background,
    theme.background
  );
  const bgHover = getAppearanceProps(
    appearance,
    backgroundHover,
    theme.backgroundHover
  );
  const bgActive = getAppearanceProps(
    appearance,
    backgroundActive,
    theme.backgroundActive
  );
  const bgFocus = getAppearanceProps(
    appearance,
    backgroundFocus,
    theme.backgroundFocus
  );
  // border
  const borderDefault = getAppearanceProps(appearance, border, theme.border);
  const bdHover = getAppearanceProps(
    appearance,
    borderHover,
    theme.borderHover
  );
  const bdActive = getAppearanceProps(
    appearance,
    borderActive,
    theme.borderActive
  );
  const bdFocus = getAppearanceProps(
    appearance,
    borderFocus,
    theme.borderFocus
  );

  const color = getAppearanceProps(appearance, text, theme.text);

  const boxShadowColor = getAppearanceProps(
    appearance,
    boxShadow,
    theme.boxShadow
  );

  /** Variable styles*/
  const fontSizeStyle = getFontSizeProps(
    elementSize,
    fontSizes,
    theme.fontSizes
  );

  const paddingStyle = getSize(padding, elementSize);
  const heightStyle = getSize(height, elementSize);

  let width;
  if (fluid) width = "100%";

  const loadingStyle = getLoadingState(props);

  const selectedStyle = getSelected(props);

  return css`
    cursor: pointer;
    outline: 0;
    border-radius: ${theme.radii ? theme.radii[1] : radii[1]};
    display: inline-block;
    text-decoration: none;
    text-align: center;
    transition: all 0.2s ease;
    user-select: none;
    vertical-align: middle;
    white-space: nowrap;
    width: ${width};
    border: 0.05rem solid;
    ${appearance === "link" && { border: "none" }};

    background: ${bgDefault};
    border-color: ${borderDefault};
    color: ${color};
    box-shadow: 0 0 0 0.1rem var(--box-shadow-color);

    &:hover {
      background: ${bgHover};
      border-color: ${bdHover};
    }

    &:active,
    &:focus {
      --box-shadow-color: ${boxShadowColor};
    }

    &:active {
      background: ${bgActive};
      border-color: ${bdActive};
    }

    &:focus {
      background: ${bgFocus};
      border-color: ${bdFocus};
    }

    &:disabled,
    &[disabled] {
      cursor: not-allowed;
      color: ${getColorFromTheme(colors, "N7", theme)} !important;
      background: ${getColorFromTheme(colors, "N4", theme)} !important;
      border: none;
    }

    font-size: ${fontSizeStyle};
    padding: ${paddingStyle};
    height: ${heightStyle};

    ${loadingStyle};
    ${selectedStyle};
  `;
};
