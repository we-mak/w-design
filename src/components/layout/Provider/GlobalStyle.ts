import { createGlobalStyle } from "styled-components";
import theme from "../../../common/styleUtils/theme";
const colors = theme.colors;

/*! normalize.css v8.0.1 | MIT License | github.com/necolas/normalize.css */
export const GlobalStyle = createGlobalStyle`
html {
  line-height: 1.15;
  -webkit-text-size-adjust: 100%;
}
body {
  margin: 0;
}
main {
  display: block;
}
h1 {
  font-size: 2em;
  margin: 0.67em 0;
}
hr {
  box-sizing: content-box;
  height: 0;
  overflow: visible;
}
pre {
  font-family: monospace, monospace;
  font-size: 1em;
}
a {
  background-color: transparent;
}
abbr[title] {
  border-bottom: none;
  text-decoration: underline;
  text-decoration: underline dotted;
}
b,
strong {
  font-weight: bolder;
}
code,
kbd,
samp {
  font-family: monospace, monospace;
  font-size: 1em;
}
small {
  font-size: 80%;
}
sub,
sup {
  font-size: 75%;
  line-height: 0;
  position: relative;
  vertical-align: baseline;
}
sub {
  bottom: -0.25em;
}
sup {
  top: -0.5em;
}
img {
  border-style: none;
}
button,
input,
optgroup,
select,
textarea {
  font-family: inherit;
  font-size: 100%;
  line-height: 1.15;
  margin: 0;
}
button,
input {
  overflow: visible;
}
button,
select {
  text-transform: none;
}
[type="button"],
[type="reset"],
[type="submit"],
button {
  -webkit-appearance: button;
}
[type="button"]::-moz-focus-inner,
[type="reset"]::-moz-focus-inner,
[type="submit"]::-moz-focus-inner,
button::-moz-focus-inner {
  border-style: none;
  padding: 0;
}
[type="button"]:-moz-focusring,
[type="reset"]:-moz-focusring,
[type="submit"]:-moz-focusring,
button:-moz-focusring {
  outline: 1px dotted ButtonText;
}
fieldset {
  padding: 0.35em 0.75em 0.625em;
}
legend {
  box-sizing: border-box;
  color: inherit;
  display: table;
  max-width: 100%;
  padding: 0;
  white-space: normal;
}
progress {
  vertical-align: baseline;
}
textarea {
  overflow: auto;
}
[type="checkbox"],
[type="radio"] {
  box-sizing: border-box;
  padding: 0;
}
[type="number"]::-webkit-inner-spin-button,
[type="number"]::-webkit-outer-spin-button {
  height: auto;
}
[type="search"] {
  -webkit-appearance: textfield;
  outline-offset: -2px;
}
[type="search"]::-webkit-search-decoration {
  -webkit-appearance: none;
}
::-webkit-file-upload-button {
  -webkit-appearance: button;
  font: inherit;
}
details {
  display: block;
}
summary {
  display: list-item;
}
template {
  display: none;
}
[hidden] {
  display: none;
}

*,
*::before,
*::after {
  box-sizing: inherit;
}

html {
  box-sizing: border-box;
  font-size: 20px;
  line-height: 1.5;
  -webkit-tap-highlight-color: transparent;
}

body {
  font-size: .8rem;
  overflow-x: hidden;
  text-rendering: optimizeLegibility;
}


h1 {
    font-size: 1.6rem;
  }

  h2 {
    font-size: 1.4rem;
  }

  h3 {
    font-size: 1.2rem;
  }

  h4 {
    font-size: 1rem;
  }
  h5 {
    font-size: 0.9rem;
  }

  h6 {
    font-size: 0.8rem;
  }

  h1,
  h2,
  h3,
  h4,
  h5,
  h6 {
    font-weight: 700;
    line-height: 1.25;
    margin-bottom: 0.5em;
    margin-top: 0;
    font-family: ${(props: any) => props.theme.fonts.medium};
  }

  h1,
  h2,
  h3,
  h4 {
    color: ${colors.N50};
  }

  h5,
  h6 {
    color: ${colors.N60};
  }

  p {
    color: ${colors.N70};
    line-height: 1.5;
    margin-bottom: 0.5em;
    margin-top: 0;
    font-family: ${(props: any) => props.theme.fonts.normal};
  }

  a {
    font-family: ${(props: any) => props.theme.fonts.normal};
    color: ${colors.B50};
    outline: none;
    text-decoration: none;

    &:hover {
      color: ${colors.B30};
    }

    &:focus,
    &:hover,
    &:active {
      text-decoration: underline;
    }
    &:focus,
    &:active {
      color: ${colors.B60};
    }
    &:focus {
      text-decoration: none;
    }
  }

  b,
  strong {
    font-weight: bolder;
  }

  small {
    font-size: 80%;
    font-weight: 400;
    color: ${colors.N30};
  }

  blockquote {
    margin-left: 0;
    padding: 0.4rem 0.8rem;
    color: ${colors.N30};
    p:last-child {
      margin-bottom: 0;
    }
  }

  table {
    border-collapse: collapse;
    border-spacing: 0;
    text-align: left;
    width: 100%;
  }
  th {
    border-bottom: 0.05rem solid ${props => props.theme.colors["N5"]};
    padding: 0.6rem 0.4rem;
    border-bottom-width: 0.1rem;
  }
  td {
    border-bottom: 0.05rem solid ${props => props.theme.colors["N2"]};
    padding: 0.6rem 0.4rem;
  }
`;
