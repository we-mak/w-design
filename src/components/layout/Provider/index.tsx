import * as React from "react";
import { ThemeProvider } from "styled-components";
import defaultTheme from "../../../common/theme";
import { ResetCss } from "./ResetCss";
import Root from "./Root";

export interface Props extends React.HTMLProps<HTMLDivElement> {
  theme?: any;
}

class Provider extends React.Component<Props> {
  stylesheet: any;

  static displayName = "WProvider";

  public static defaultProps: Props = {
    theme: {}
  };

  componentDidMount() {
    this.stylesheet = document.createElement("style");
    this.stylesheet.type = "text/css";
    // this.stylesheet.setAttribute("rel", "preload");
    // this.stylesheet.setAttribute("as", "style");
    this.stylesheet.innerHTML = ResetCss;

    if (document && document.head) {
      document.head.appendChild(this.stylesheet);
    }
  }

  componentWillUnmount() {
    if (this.stylesheet && document && document.head) {
      document.head.removeChild(this.stylesheet);
      delete this.stylesheet;
    }
  }

  render() {
    const { theme, ...props } = this.props;
    return (
      <ThemeProvider theme={{ ...defaultTheme, ...theme }}>
        <Root {...props as any} />
      </ThemeProvider>
    );
  }
}

export default Provider;
