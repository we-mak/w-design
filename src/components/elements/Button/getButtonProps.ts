import { BtnProps } from "./buttonType";

const getAppearanceProps = (props: BtnProps) => {
  const {
    appearance,
    elementSize,
    fluid,
    className,
    isDisabled,
    isLoading
  } = props;

  return {
    appearance,
    elementSize,
    fluid,
    className,
    disabled: isDisabled,
    isLoading
  };
};

const getLinkElementProps = (props: BtnProps) => {
  const { href, target } = props;
  return { href, target };
};

const getButtonElementProps = (props: BtnProps) => {
  const { ariaHaspopup, ariaExpanded, ariaControls, form, type } = props;

  return {
    "aria-haspopup": ariaHaspopup,
    "aria-expanded": ariaExpanded,
    "aria-controls": ariaControls,
    "aria-label": props.ariaLabel,
    form,
    type,
    id: props.id
  };
};

const getEventProps = (component: any) => {
  const { onBlur, onFocus } = component;
  const { onClick, tabIndex } = component.props;

  return {
    onBlur,
    onClick,
    onFocus,
    tabIndex
  };
};

export const getButtonProps = (component: any) => {
  const { props } = component;

  const defaultProps = {
    ...getAppearanceProps(props),
    ...getEventProps(component)
  };

  if (props.href) {
    if (props.isDisabled) return defaultProps;

    return {
      ...defaultProps,
      ...getLinkElementProps(props)
    };
  }

  return {
    ...defaultProps,
    ...getButtonElementProps(props)
  };
};
