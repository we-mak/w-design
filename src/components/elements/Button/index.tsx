/**
 * Button
 *
 */
import * as React from "react";
import styled, { AnyStyledComponent } from "styled-components";
import { ButtonProps } from "./types";
import { getButtonStyle } from "./getButtonStyle";
import Icon from "../Icon";

const StyledButton = styled.button`
  ${props => getButtonStyle(props)};
`;

StyledButton.displayName = "StyledButton";

const StyledA = styled.a`
  a& {
    line-height: initial;
    &:hover {
      text-decoration: none;
    }
    ${props => getButtonStyle(props)};
  }
`;

StyledA.displayName = "StyledAButton";

const StyledSpan = styled.span`
  ${props => getButtonStyle(props)};
`;
StyledSpan.displayName = "StyledSpanButton";

// TODO: add auto focus style

const Button = (props: ButtonProps) => {
  const {
    buttonRef,
    appearance = "default",
    type = "button",
    size = "md",
    className,
    id,
    fluid,
    iconBefore,
    iconAfter,
    iconOnly,
    ariaHaspopup,
    ariaExpanded,
    ariaControls,
    ariaLabel,
    form,
    isLoading = false,
    isDisabled = false,
    isSelected = false,
    href,
    target,
    autoFocus,
    onClick,
    tabIndex,
    children,
    ...rest
  } = props;

  let ButtonComponent: AnyStyledComponent;

  if (href) {
    isDisabled ? (ButtonComponent = StyledSpan) : (ButtonComponent = StyledA);
  } else {
    ButtonComponent = StyledButton;
  }

  let button!: HTMLElement | undefined;
  const getButtonRef = (ref: HTMLElement) => {
    button = ref;
    if (buttonRef) buttonRef(ref);
  };

  React.useEffect(() => {
    if (autoFocus && button) button.focus();
    return () => (button = undefined);
  });

  return (
    <ButtonComponent
      ref={getButtonRef}
      //Appearance Props
      appearance={appearance}
      size={size}
      fluid={fluid}
      className={className}
      id={id}
      iconOnly={iconOnly}
      disabled={isDisabled}
      // TODO: change this to state
      isLoading={isLoading}
      isSelected={isSelected}
      // Link Props
      href={href}
      target={target}
      // Button Props
      type={type}
      form={form}
      aria-haspopup={ariaHaspopup}
      aria-expanded={ariaExpanded}
      aria-controls={ariaControls}
      aria-label={ariaLabel}
      tabIndex={tabIndex}
      onClick={isDisabled ? undefined : onClick}
      {...rest}
    >
      {iconBefore && <Icon className={`icon-before ${iconBefore}`} />}
      {children}
      {iconAfter && <Icon className={`icon-after ${iconAfter}`} />}
    </ButtonComponent>
  );
};

export default Button;
