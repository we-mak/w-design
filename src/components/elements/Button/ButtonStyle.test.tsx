import * as React from "react";
import * as renderer from "react-test-renderer";
import { mount } from "enzyme";
import "jest-styled-components";
import Button from "./Button";
import Provider from "../../layout/Provider/Provider";

describe("test Button style", () => {
  // render test
  it("should render right style", () => {
    const tree = renderer.create(<Button />).toJSON();
    expect(tree).toMatchSnapshot();
    const fullTree = renderer
      .create(
        <Provider>
          <Button>Test</Button>
        </Provider>
      )
      .toJSON();
    expect(fullTree).toMatchSnapshot();
  });

  // get style test
  it("should renders correctly with theme provider", () => {
    const defaultBtn = mount(
      <Provider>
        <Button>Test</Button>
      </Provider>
    );
    expect(defaultBtn.find("button")).toHaveStyleRule("background", "#E2EAF2");
    const primaryBtn = mount(
      <Provider>
        <Button appearance="primary">Test</Button>
      </Provider>
    );
    expect(primaryBtn.find("button")).toHaveStyleRule("background", "#008CC0");
  });

  // appearance
  it("should has right appearance style", () => {
    const defaultBtn = mount(<Button />);
    const primaryBtn = mount(<Button appearance="primary" />);
    const dangerBtn = mount(<Button appearance="danger" />);
    const warnBtn = mount(<Button appearance="warning" />);
    const successBtn = mount(<Button appearance="success" />);
    expect(defaultBtn).toHaveStyleRule("background", "#E2EAF2");
    expect(primaryBtn).toHaveStyleRule("background", "#008CC0");
    expect(dangerBtn).toHaveStyleRule("background", "#DC1F1F");
    expect(warnBtn).toHaveStyleRule("background", "#FA9B00");
    expect(successBtn).toHaveStyleRule("background", "#29A634");
  });
  // button size
  it("should has right size style", () => {
    const smBtn = mount(<Button elementSize="sm" />);
    const defaultBtn = mount(<Button />);
    const lgBtn = mount(<Button elementSize="lg" />);
    expect(smBtn).toHaveStyleRule("padding", "0.15rem 0.3rem");
    expect(smBtn).toHaveStyleRule("font-size", "0.7rem");
    expect(defaultBtn).toHaveStyleRule("padding", "0.35rem 0.4rem");
    expect(defaultBtn).toHaveStyleRule("font-size", "0.8rem");
    expect(lgBtn).toHaveStyleRule("padding", "0.45rem 0.6rem");
    expect(lgBtn).toHaveStyleRule("font-size", "0.9rem");
  });
  // loading style
  it("should render loading style", () => {
    const btn = mount(<Button isLoading />);
    expect(btn).toHaveStyleRule("animation", "loading 500ms infinite linear", {
      modifier: "::after"
    });

    const primaryBtn = mount(<Button isLoading appearance="primary" />);
    expect(primaryBtn).toHaveStyleRule("border-color", "#FFF", {
      modifier: "::after"
    });
  });
  // selected style
  it("should render selected style", () => {
    const btn = mount(<Button isSelected />);
    expect(btn).toHaveStyleRule("background", "#004660");

    const warnBtn = mount(<Button isSelected appearance="warning" />);
    expect(warnBtn).toHaveStyleRule("background", "#EF8100");
    expect(warnBtn).toHaveStyleRule("background", "#EF8100", {
      modifier: ":hover"
    });

    const dangerBtn = mount(<Button isSelected appearance="danger" />);
    expect(dangerBtn).toHaveStyleRule("background", "#DC1F1F");
  });
  // fluid button
  it("should render fluid style", () => {
    const btn = mount(<Button fluid />);
    expect(btn).toHaveStyleRule("width", "100%");
  });
});
