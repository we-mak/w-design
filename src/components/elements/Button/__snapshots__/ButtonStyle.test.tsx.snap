// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`test Button style should render right style 1`] = `
.c0 {
  cursor: pointer;
  outline: 0;
  border-radius: 0.1rem;
  display: inline-block;
  -webkit-text-decoration: none;
  text-decoration: none;
  text-align: center;
  -webkit-transition: all 0.2s ease;
  transition: all 0.2s ease;
  -webkit-user-select: none;
  -moz-user-select: none;
  -ms-user-select: none;
  user-select: none;
  vertical-align: middle;
  white-space: nowrap;
  border: 0.05rem solid;
  background: #E2EAF2;
  border-color: #E2EAF2;
  color: #10161A;
  box-shadow: 0 0 0 0.1rem var(--box-shadow-color);
  font-size: 0.8rem;
  padding: 0.35rem 0.4rem;
  height: 1.8rem;
}

.c0:hover {
  background: #CED9E3;
  border-color: #CED9E3;
}

.c0:active,
.c0:focus {
  --box-shadow-color: transparent;
}

.c0:active {
  background: #B1EAFF;
  border-color: #B1EAFF;
}

.c0:focus {
  background: #B1EAFF;
  border-color: #B1EAFF;
}

.c0:disabled,
.c0[disabled] {
  cursor: not-allowed;
  color: #90A9BF !important;
  background: #CED9E3 !important;
  border: none;
}

<button
  aria-controls={undefined}
  aria-expanded={undefined}
  aria-haspopup={undefined}
  aria-label={undefined}
  className="c0"
  disabled={false}
  form={undefined}
  id={undefined}
  onBlur={[Function]}
  onClick={undefined}
  onFocus={[Function]}
  tabIndex={undefined}
  type="button"
/>
`;

exports[`test Button style should render right style 2`] = `
.c1 {
  cursor: pointer;
  outline: 0;
  border-radius: 0.1rem;
  display: inline-block;
  -webkit-text-decoration: none;
  text-decoration: none;
  text-align: center;
  -webkit-transition: all 0.2s ease;
  transition: all 0.2s ease;
  -webkit-user-select: none;
  -moz-user-select: none;
  -ms-user-select: none;
  user-select: none;
  vertical-align: middle;
  white-space: nowrap;
  border: 0.05rem solid;
  background: #E2EAF2;
  border-color: #E2EAF2;
  color: #10161A;
  box-shadow: 0 0 0 0.1rem var(--box-shadow-color);
  font-size: 0.8rem;
  padding: 0.35rem 0.4rem;
  height: 1.8rem;
}

.c1:hover {
  background: #CED9E3;
  border-color: #CED9E3;
}

.c1:active,
.c1:focus {
  --box-shadow-color: transparent;
}

.c1:active {
  background: #B1EAFF;
  border-color: #B1EAFF;
}

.c1:focus {
  background: #B1EAFF;
  border-color: #B1EAFF;
}

.c1:disabled,
.c1[disabled] {
  cursor: not-allowed;
  color: #90A9BF !important;
  background: #CED9E3 !important;
  border: none;
}

.c0 {
  font-family: -apple-system,BlinkMacSystemFont,Helvetica,"Helvetica Neue",system-ui,sans-serif;
}

.c0 h1 {
  font-size: 2rem;
}

.c0 h2 {
  font-size: 1.6rem;
}

.c0 h3 {
  font-size: 1.4rem;
}

.c0 h4 {
  font-size: 1.2rem;
}

.c0 h5 {
  font-size: 1rem;
}

.c0 h6 {
  font-size: 0.8rem;
}

.c0 h1,
.c0 h2,
.c0 h3,
.c0 h4,
.c0 h5,
.c0 h6 {
  font-weight: 600;
  line-height: 1.25;
  margin-bottom: 0.5em;
  margin-top: 0;
}

.c0 h1,
.c0 h2,
.c0 h3,
.c0 h4 {
  color: #10161A;
}

.c0 h5,
.c0 h6 {
  color: #182026;
}

.c0 p {
  color: #202B33;
  line-height: 1.5;
  margin-bottom: 0.5em;
  margin-top: 0;
}

.c0 a {
  color: #008CC0;
  outline: none;
  -webkit-text-decoration: none;
  text-decoration: none;
}

.c0 a:hover {
  color: #00A7E4;
}

.c0 a:focus,
.c0 a:hover,
.c0 a:active {
  -webkit-text-decoration: underline;
  text-decoration: underline;
}

.c0 a:focus,
.c0 a:active {
  color: #007099;
}

.c0 a:focus {
  -webkit-text-decoration: none;
  text-decoration: none;
}

.c0 b,
.c0 strong {
  font-weight: bolder;
}

.c0 code,
.c0 kbd,
.c0 pre,
.c0 samp {
  font-family: "Menlo, Courier, monospace";
}

.c0 code {
  color: #F22828;
  background: #FBFBFB;
  padding: 0.1rem 0.15rem;
  line-height: 1.2;
}

.c0 small {
  font-size: 80%;
  font-weight: 400;
  color: #566E82;
}

.c0 blockquote {
  margin-left: 0;
  padding: 0.4rem 0.8rem;
  color: #566E82;
}

.c0 blockquote p:last-child {
  margin-bottom: 0;
}

<div
  className="c0"
>
  <button
    aria-controls={undefined}
    aria-expanded={undefined}
    aria-haspopup={undefined}
    aria-label={undefined}
    className="c1"
    disabled={false}
    form={undefined}
    id={undefined}
    onBlur={[Function]}
    onClick={undefined}
    onFocus={[Function]}
    tabIndex={undefined}
    type="button"
  >
    Test
  </button>
</div>
`;
